src/Featurizers/Components/VectorNormsEstimator.h

```diff
@@ -0,0 +1,181 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+#include <limits>
+
+#include "TrainingOnlyEstimatorImpl.h"
+#include "../../Traits.h"
+#include "NormUpdaters.h"
+#include "../Structs.h"
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+namespace Components {
+
+static constexpr char const * const         VectorNormsEstimatorName("VectorNormsEstimator");
+
+namespace TypeSelector {
+
+// updater type selector is created to select different updater based on input parameter
+template <typename T, NormType NormT>
+struct UpdaterTypeSelector {
+};
+template <typename T>
+struct UpdaterTypeSelector<T, NormType::L1> {
+    using type = Updaters::L1NormUpdater<T>;
+};
+
+template <typename T>
+struct UpdaterTypeSelector<T, NormType::L2> {
+    using type = Updaters::L2NormUpdater<T>;
+};
+
+template <typename T>
+struct UpdaterTypeSelector<T, NormType::MAX> {
+    using type = Updaters::MaxNormUpdater<T>;
+};
+
+} // namespace TypeSelector
+
+
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         VectorNormsAnnotationData
+///  \brief         An annotation class which contains the norms of a matrix
+///
+class VectorNormsAnnotationData {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Data
+    // |
+    // ----------------------------------------------------------------------
+    std::vector<std::double_t> const Norms;
+    NormType                 const Type;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    VectorNormsAnnotationData(std::vector<std::double_t> norms, NormType type);
+    ~VectorNormsAnnotationData(void) = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(VectorNormsAnnotationData);
+};
+
+namespace Details {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         VectorNormsTrainingOnlyPolicy
+///  \brief         `VectorNormsEstimator` implementation details.
+///
+template <typename T, NormType NormT>
+class VectorNormsTrainingOnlyPolicy {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    static_assert(Traits<T>::IsVectorType, "Input type has to be vector!");

```

Comment:
>static_assert(Traits<T>::IsVectorType, "Input type has to be vector!"); [](start = 4, length = 71)

We had discussed using iterators representing a range rather than iterator types; what was the result of that conversation?